[linux-port] Add missing case statements

pow2clk opened the pull request:
When switching over an enum value, the ideal is to include a case
for every possibly value. Where possible, that's what this change
does to silence warnings and also to provide space to properly
handle every possible input and throw a warning should a new one
be added. In some cases, it's not practical to list every input.
In these cases, putting a default that doesn't change the behavior,
but let's the compiler know that the behavior is intended.
Fixes 133 clang and 8108(!) gcc warnings.

Contributes to https://github.com/google/DirectXShaderCompiler/issues/206

ehsannas commented on lib/HLSL/DxilDebugInstrumentation.cpp:
```diff
@@ -691,6 +691,9 @@ void DxilDebugInstrumentation::addStepDebugEntry(BuilderContext &BC, Instruction
   case Type::TypeID::FunctionTyID:
   case Type::TypeID::ArrayTyID:
   case Type::TypeID::VectorTyID:
+  case Type::TypeID::X86_FP80TyID:
```
This is a change in behavior. Is it the right behavior? I don't know...

ehsannas commented on lib/HLSL/DxilShaderAccessTracking.cpp:
```diff
@@ -180,8 +180,12 @@ static char EncodeRegisterType(RegisterType r) {
   case RegisterType::CBV:     return 'C';
   case RegisterType::SRV:     return 'S';
   case RegisterType::UAV:     return 'U';
+  case RegisterType::DSV:     return 'D';
```
Changes behavior .... It does look like a reasonable change though.

ehsannas commented on tools/clang/lib/AST/ASTContext.cpp:
```diff
@@ -4619,6 +4619,8 @@ QualType ASTContext::getFloatingTypeOfSizeWithinDomain(QualType Size,
     case FloatRank:      return FloatComplexTy;
     case DoubleRank:     return DoubleComplexTy;
     case LongDoubleRank: return LongDoubleComplexTy;
+    case LitFloatRank:   llvm_unreachable("Complex LitFloat half is not supported");  // HLSL Change
```
`Complex LitFloat is not supported`

ehsannas commented on tools/clang/lib/AST/ASTContextHLSL.cpp:
```diff
@@ -1140,14 +1140,12 @@ UnusualAnnotation* hlsl::UnusualAnnotation::CopyToASTContext(ASTContext& Context
   for (const UnusualAnnotation *it : decl->getUnusualAnnotations()) {
-    switch (it->getKind()) {
-    case UnusualAnnotation::UA_SemanticDecl: {
+    if (it->getKind() == UnusualAnnotation::UA_SemanticDecl) {
```
haha.. Nice

ehsannas commented on tools/clang/lib/CodeGen/CGExprConstant.cpp:
```diff
@@ -769,9 +769,14 @@ class ConstExprEmitter :
           cast<llvm::ArrayType>(ST->getElementType(0)), Rows);
       return llvm::ConstantStruct::get(ST, Mat);
     }
+    case CK_HLSLVectorToScalarCast:
```
do we need these 3 case statements now that we have added the default below?

ehsannas commented on tools/clang/lib/Sema/SemaOverload.cpp:
```diff
@@ -5109,7 +5109,14 @@ static bool CheckConvertedConstantConversions(Sema &S,
   case ICK_Zero_Event_Conversion:
-  case ICK_Flat_Conversion: // HLSL Change
+ // HLSL Change Start
+  case ICK_Flat_Conversion:
```
change in behavior: These cases used to hit `llvm_unreachable`, but now hit `false`. Is this the correct behavior?

ehsannas commented on tools/clang/lib/Sema/DeclSpec.cpp:
```diff
@@ -309,6 +309,10 @@ bool Declarator::isDeclarationOfFunction() const {
     case TST_void:
     case TST_wchar:
+    // HLSL Change Starts
+    case TST_min10float:
```
change in behavior: These cases used to hit llvm_unreachable, but now hit false. Is this the correct behavior?

pow2clk replied:
these types can't be function declarations. So returning false seems correct.

ehsannas commented on tools/clang/lib/CodeGen/CGDebugInfo.cpp:
```diff
@@ -449,6 +449,26 @@ llvm::DIType *CGDebugInfo::CreateType(const BuiltinType *BT) {
+  case BuiltinType::LitInt:
+  case BuiltinType::Min10Float:
+  case BuiltinType::LitFloat:
+    llvm_unreachable("Unsupported HLSL types");
```
Not sure about this one... they should probably actually implement these cases? >.<

pow2clk replied:
This function is for generating DWARF data for debugging purposes. I don't think any HLSL debuggers use that format. Do they?

ehsannas replied:
I guess not.

pow2clk commented:
These are the tricky ones. Every bug I ran into developing this was due to this kind of changed behavior. I'll look at these more closely and maybe split them up further to better drill down which can safely be resolved with a default: break;, which should be errors, and which should have actual implementations.

pow2clk commented:
I've audited all the case and default statements I added. I've added comments explaining the default statements. I removed any that were less certain so they can be addressed in a later change with more scrutiny.

pow2clk commented on tools/clang/lib/Sema/SemaHLSL.cpp:
```diff
@@ -7314,21 +7392,8 @@ bool HLSLExternalSource::CanConvert(
-        if (explicitConversion) {
-          Second = ICK_Flat_Conversion;
-          goto lSuccess;
-        }
```
This conditional and switch is a duplicate of the previous. It depends on the same variables for the result and changes none of them. So the repetition accomplishes nothing.

ehsannas replied:
😲

ehsannas replied:
Good catch!

ehsannas commented on tools/clang/lib/Sema/SemaHLSL.cpp:
```diff
@@ -4665,9 +4670,10 @@ static bool CombineObjectTypes(ArBasicKind Target, _In_ ArBasicKind Source,
+    AssignOpt(AR_BASIC_UNKNOWN, pCombined);
   }
-
-  AssignOpt(AR_BASIC_UNKNOWN, pCombined);
```
I think in some cases (where the case matches, but the if condition inside the case isn't true), AssignOpt used to be executed, but now won't.

pow2clk replied:
Yeah. it looks like you're right. I'll move it back out and make it a break.

ehsannas commented:
Bots passing.
